### DISCUSSION TITLE
Strange dynamic namespace and empty default one

### DISCUSSION THREAD

**Discussion by hfhbd** [pdvrieze/xmlutil#196] [General]
When encoding my classes to a String, sometimes I get this output:
```xml
<zdef-264009233:ReceivingSystem xmlns="" xmlns:zdef-264009233="http://myrealns.example.com/foo" listID="null" listVersionID="null" name="null" languageID="null">ABC</zdef-264009233:ReceivingSystem>
```

```kotlin
@XmlElement
@XmlSerialName(
  value = "ReceivingSystem",
  namespace = "http://myrealns.example.com/foo",
)
val receivingSystem: Code? = null,


@Serializable
@XmlSerialName(
  value = "Code",
  namespace = "http://myrealns.example.com/bar",
)
data class Code(
  @XmlValue
  val `value`: String,
  al listID: String? = null,
  val listVersionID: String? = null,
  val name: String? = null,
  val languageID: LanguageTagCode? = null,
)
```

This happens with different classes, small as well big ones but I am unable to create a small reproducer...

While this is valid xml, it makes testing harder. Do you have any idea how this behavior is triggered? I didn't find the `zdef` prefix in the code of xmlutil and I am lost.

Kotlin 1.9.22, serialization 1.6.2, xmutil 0.86.3

**Comment by pdvrieze** [comment]
The code does generate prefixes if it doesn't have a an existing one, but they start with ns1 and increment until an unused prefix is found. However, by default serialization uses the "native" serializer (as in the stax parser/serializer provided by the jvm (and this can be adjusted using service loaders - at jvm level). I would suspect that your system loads such an xml library (which gets picked up automatically - I only test against the default jvm library and woodstox), and that this library generates crappy prefixes.

This doesn't mean however that there isn't a bug. I'll have a check with the default implementation.

**Comment by hfhbd** [comment]
I will try to isolate a reproducer, AFAIK some `ID` and `Version` named members caused the bug, sometimes.

**Comment by pdvrieze** [comment]
Your reproducer has some gaps (null attributes will actually mean empty attributes, but I have managed to reproduce the output with the following reproducer.

```kotlin
/**
 * Tests for but #196 with strange namespaces.
 */
class StrangeNamespace196 {

    @Test
    fun testSerialize() {
        val xml = XML {
            recommended_0_87_0 {
            }
        }

//        val data = Container(Code("ABC", "null", "null", "null", LanguageTagCode("null")))
        val data = Container(Code("ABC", "null", "null", "null", LanguageTagCode("en") ))

        val serialized = xml.encodeToString(Container.serializer(), data)
        assertXmlEquals("<Container></Container>", serialized)

    }

    @Serializable
    class Container(
    @XmlElement
    @XmlSerialName(
        value = "ReceivingSystem",
        namespace = "http://myrealns.example.com/foo",
    )
    val receivingSystem: Code? = null)


    @Serializable
    @XmlSerialName(
        value = "Code",
        namespace = "http://myrealns.example.com/bar",
    )
    data class Code(
        @XmlValue
        val `value`: String,
        val listID: String? = null,
        val listVersionID: String? = null,
        val name: String? = null,
        val languageID: LanguageTagCode? = null,
    )

    @Serializable
    @JvmInline
    value class LanguageTagCode(val lang: String)

}
```

**Reply by hfhbd** [reply]
Thank you for looking into it, really appreciate it! 
Do you have any idea what's causing the bug or how to prevent it? 

I adjusted my code generator to switch from value classes to normal data classes, btw, but this doesn't fix it: 
```kotlin
@Serializable
@XmlSerialName(
  value = "Code",
  namespace = "http://myrealns.example.com/bar",
)
data class Code(
  @XmlValue
  val `value`: String,
  val listID: String? = null,
  val listVersionID: String? = null,
  val name: String? = null,
  val languageID: String? = null,
)
```

**Reply by hfhbd** [reply]
Hm, if I either remove the namespace from the `Container` or remove `b: String` from `System`, then I don't get the namespace issue:
```kotlin
@Serializable
@XmlSerialName(value = "Container", namespace = "http://myrealns.example.com/a")
class Container(
  @XmlElement
  @XmlSerialName(value = "ReceivingSystem", namespace = "http://myrealns.example.com/foo")
  val receivingSystem: System
)

@Serializable
@XmlSerialName(value = "System", namespace = "http://myrealns.example.com/bar")
data class System(
  val a: String,
  val b: String,
)
```
Unfortunately, I can't do it because I need all members and I also use the class `Container` as top-level as well as member of other classes.

**Comment by pdvrieze** [comment]
Ok. I've figured out what is going on. This is a problem with the default xml writer provided by the JVM (11 all the way to 21) when it is in namespace repairing mode (the default - one of the config flags).

**"Fix":**
```kotlin
val xml = XML { repairNamespaces=false }
```

The serialization module is designed to write namespace attributes when needed (and doesn't rely on the "automatic repair mode" )

The short of it: The default JVM serializer is somewhat "broken" (in this case in namespace repairing mode)

The long, what happens is: 
 - omitting the prefix for the type (with "foo" namespace) this defaults to the default (empty) prefix.
 - Attributes without namespace must be in the default prefix (and for attributes an empty prefix always maps to the default/null namespace)
 - This means there is a potential conflict of namespace prefixes (however this only happens if you want to write an attribute in the "foo" namespace)
 - In repairing mode the serializer (Java, not xmlutil) is pessimistic and doesn't allow this. It forces the default prefix to the default namespace (not needed at all), and generates a new (really ugly) namespace for the "foo" namespace. (there actually appears to be a bug with the "last" attribute)

**Reply by hfhbd** [reply]
Wow, thank you for the quick analysis and the workaround!

**Reply by pdvrieze** [reply]
@hfhbd You're welcome. While the generated (repaired) XML is valid it is ugly (and the repair code is "buggy" - it "corrects" attributes, but treats the last one differently, hence it works if there is only one attribute, but triggers the ugliness with more than one.
